core/src/types/read/reader.rs

```diff
@@ -53,13 +60,50 @@ impl Reader {
     /// in crate only.
     pub(crate) async fn create(
         acc: Accessor,
-        path: &str,
+        path: Arc<String>,
         args: OpRead,
         options: OpReader,
     ) -> Result<Self> {
-        let (_, r) = acc.read(path, args).await?;
+        let (_, r) = acc.read(&path, args).await?;
+
+        Ok(Reader {
+            acc,
+            path,
+            inner: r,
+            options,
+            size: Arc::new(AtomicContentLength::new()),
+        })
+    }
+
+    /// Parse users input range bounds into valid `Range<u64>`.
+    ///
+    /// To avoid duplicated stat call, we will cache the size of the reader.
+    async fn parse_range(&self, range: impl RangeBounds<u64>) -> Result<Range<u64>> {
+        let start = match range.start_bound() {
+            Bound::Included(v) => *v,
+            Bound::Excluded(v) => v + 1,
+            Bound::Unbounded => 0,
+        };
+
+        let end = match range.end_bound() {
+            Bound::Included(v) => v + 1,
+            Bound::Excluded(v) => *v,
+            Bound::Unbounded => match self.size.load() {
+                Some(v) => v,
+                None => {
+                    let size = self
+                        .acc
+                        .stat(&self.path, OpStat::new())
+                        .await?
+                        .into_metadata()
+                        .content_length();
+                    self.size.store(size);
+                    size
+                }
+            },
```

Comment:
Does this default behavior need to be updated in the documentation?
(Automatically limits to the actual content length if the end is not given.)

Reply:
The actual behavior doesn't changed.